src/Runtime/Distributions/Automata/IWeightFunction.cs

```diff
@@ -0,0 +1,325 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.ML.Probabilistic.Distributions.Automata
+{
+    using Microsoft.ML.Probabilistic.Factors.Attributes;
+    using Microsoft.ML.Probabilistic.Math;
+    using System;
+    using System.Collections.Generic;
+    using System.Text;
+
+    /// <summary>
+    /// The part of <see cref="IWeightFunction{TSequence, TElement, TElementDistribution, TSequenceManipulator, TAutomaton, TThis}"/> API,
+    /// that is agnostic of the concrete weight function class. Implementations of closed constructed
+    /// <see cref="IWeightFunction{TSequence, TElement, TElementDistribution, TSequenceManipulator, TAutomaton, TThis}"/>
+    /// with the only differing type parameter being TThis, can all be upcasted to this type.
+    /// </summary>
+    /// <typeparam name="TSequence">The type of a sequence.</typeparam>
+    /// <typeparam name="TElement">The type of a sequence element.</typeparam>
+    /// <typeparam name="TElementDistribution">The type of a distribution over sequence elements.</typeparam>
+    /// <typeparam name="TSequenceManipulator">The type providing ways to manipulate sequences.</typeparam>
+    /// <typeparam name="TAutomaton">The type of a weighted finite state automaton, that can be used to
+    /// represent all weight functions.</typeparam>
+    [Quality(QualityBand.Experimental)]
+    public interface IWeightFunction<TSequence, TElement, TElementDistribution, TSequenceManipulator, TAutomaton>
+        where TSequence : class, IEnumerable<TElement>
+        where TElementDistribution : IDistribution<TElement>, SettableToProduct<TElementDistribution>, SettableToWeightedSumExact<TElementDistribution>, CanGetLogAverageOf<TElementDistribution>, SettableToPartialUniform<TElementDistribution>, new()
+        where TSequenceManipulator : ISequenceManipulator<TSequence, TElement>, new()
+        where TAutomaton : Automaton<TSequence, TElement, TElementDistribution, TSequenceManipulator, TAutomaton>, new()
+    {
+        /// <summary>
+        /// Returns an automaton representation of the current weight function. If the weight function is normalized,
+        /// the resulting automaton is stochastic, i.e. the sum of weights of all the outgoing transitions and the ending weight is 1 for every node.
+        /// </summary>
+        TAutomaton AsAutomaton();
+
+        /// <summary>
+        /// Gets the only point to which the current function assigns a non-zero weight.
+        /// </summary>
+        /// <exception cref="System.InvalidOperationException">Thrown when an attempt is made to get the <see cref="Point"/>
+        /// of a non-pointmass weight function.</exception>
+        TSequence Point { get; }
```

Comment:
If it is expensive to compute Point when your distribution is a point mass, then your class is badly designed.

Reply:
I disagree. If you have an automata, then computing point over it is an expensive operation that involves full traversal and a lot of bookkeeping.

Obviously our `SequenceDistribution` type tries to switch back to point representation when it can. But the Point for automata representation remains an expensive operation.

Reply:
You say "I disagree" then you say that Point is currently expensive for automata.  I don't see how your second statement is a disagreement with my statement.

Reply:
So, I thought, and moving the responsibility for calling NormalizeStructure from SequenceDistribution to MultiRepresentableWeightFunction is a topic for a separate PR.

About the initial comment: long published IDistribution<T> interface defines Point and IsPointMass properties and we're not going to change it here. Weight functions exist in our codebase to support the needs of SequenceDistribution, so, I think, it makes sense to leave them separate in IWeightFunction as well.